src/interpolate.rs

```diff
@@ -0,0 +1,249 @@
+use {Duplex, Frame, Sample};
+
+use core::marker::PhantomData;
+
+/// An iterator that converts the rate at which frames are yielded from some given frame
+/// Interpolator into a new type.
+///
+/// Other names for `sample::rate::Converter` might include:
+///
+/// - Sample rate converter
+/// - {Up/Down}sampler
+/// - Sample interpolater
+/// - Sample decimator
+///
+#[derive(Clone)]
+pub struct Converter<T: Iterator, I: Interpolator<T>>
+    where <T as Iterator>::Item: Frame
+{
+    interpolator: I,
+    interpolation_value: f64,
+    source_to_target_ratio: f64,
+    phantom: PhantomData<T>
+}
+
+/// Interpolator that just rounds off any values to the previous value from the source
+pub struct Floor<I>
+    where I: Iterator,
+          I::Item: Frame
+{
+    source: I,
+    left: I::Item 
+}
+
+/// Interpolator that interpolates linearly between the previous value and the next value
+pub struct Linear<I>
+    where I: Iterator,
+          I::Item: Frame
+{
+    source: I,
+    left: I::Item,
+    right: Option<I::Item>
+}
+
+/// Trait for all things that can interpolate between two values. Implementations should keep track
+/// of the necessary data both before and after the current frame.
+pub trait Interpolator<I>
```

Comment:
I don't think `Interpolator` really needs this `I` parameter - A single associated `Frame` type (yielded by the `fn interpolate` method) should be enough.

This way implementations might look more like this:
```rust
impl<I> Interpolator for Linear<I>
    where I: Iterator,
          I::Item: Frame,
          <I::Item as Frame>::Sample: Duplex<f64>,
{
    type Frame = I::Item;
    ...
}
```

Which should allow us to remove the `T` type parameter from `Converter`